### PR TITLE
fix: #2642 root element have no __typename field

### LINKED ISSUE #2642: urql 3.0 - regression on maskTypename
Hi! 👋 
      
Firstly, thanks for your work on this project! 🙂

Today I used [patch-package](https://github.com/ds300/patch-package) to patch `@urql/core@3.0.1` for the project I'm working on.

With urql@3.0, even if `maskTypename` option is set, `__typename` fields are not removed from response.

Here is the diff that solved my problem:

```diff
diff --git a/node_modules/@urql/core/dist/urql-core.js b/node_modules/@urql/core/dist/urql-core.js
index 0724265..eef71f3 100644
--- a/node_modules/@urql/core/dist/urql-core.js
+++ b/node_modules/@urql/core/dist/urql-core.js
@@ -74,11 +74,12 @@ var formatDocument = node => {
 };
 
 var maskTypename = data => {
+ 
   if (!data || typeof data !== 'object') {
     return data;
   } else if (Array.isArray(data)) {
     return data.map(maskTypename);
-  } else if (data && typeof data === 'object' && '__typename' in data) {
+  } else if (data && typeof data === 'object') {
     var acc = {};
 
     for (var key in data) {
@@ -791,6 +792,7 @@ var Client = function Client(opts) {
       }))(wonka.take(1)(wonka.filter(op => op.kind === 'query' && op.key === operation.key && op.context.requestPolicy !== 'cache-only')(operations$)))]);
     })( // End the results stream when an active teardown event is sent
     wonka.takeUntil(wonka.filter(op => op.kind === 'teardown' && op.key === operation.key)(operations$))(result$)))));
+ 
     return source;
   };
 
```

<em>This issue body was [partially generated by patch-package](https://github.com/ds300/patch-package/issues/296).</em>

### PR DESCRIPTION
**Pull request by mlecoq**
Resolves #2642

### CODE REVIEW

File: .changeset/many-worms-bow.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@urql/core': minor
+---
+
+fix maskTypename
```
Comment (JoviDeCroock) [inline .changeset/many-worms-bow.md:5]:
```suggestion
Fix case where `maskTypename` would not traverse down when the root query-field does not contain a `__typename`
```

**Review by JoviDeCroock — COMMENTED** [review]
Mind adding a test for this?

File: .changeset/many-worms-bow.md (hunk 1 above)
Comment (mlecoq) [reply .changeset/many-worms-bow.md]: Thanks for your feedback ! 

I have added a unit test and fixed the case with Date objects

**Comment by mlecoq** [comment]
> Mind adding a test for this?

Hi, does it look good to you after last changes ?

**Comment by kitten** [comment]
Hiya 👋 Sorry about that regression. That was my bad

Unfortunately, the fix here you've got now has a small problem as well.
You've got the `Date` instance check there, however, exchanges may add any kind of objects into the result, making the recursion now potentially dangerous and destroying instances or even recursing infinitely if instances with recursive properties are added to a result 😅

What I forgot when refactoring is that we skip adding a `__typename` field to the root selection set. However, what we can do about this is skip the `__typename` check for the first run of this function, which we could do with an additional argument for instance

**Comment by mlecoq** [comment]
Ok I'll change it

**Review by JoviDeCroock — APPROVED** [review]

**Pull request merged**